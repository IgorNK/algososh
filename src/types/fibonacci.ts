export interface IFibonacciSequenceHandler {
  calculate: () => Promise<void>;
}
