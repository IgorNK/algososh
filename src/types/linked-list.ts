export interface ILinkedListDisplayHandler {
  addHead: (item: string) => Promise<void>;
  addTail: (item: string) => Promise<void>;
  removeHead: () => Promise<void>;
  removeTail: () => Promise<void>;
  addAt: (item: string, index: number) => Promise<void>;
  removeAt: (index: number) => Promise<void>;
}

interface INode<T> {
  next: Node<T> | null;
  prev: Node<T> | null;
}

export class Node<T> implements INode<T> {
  public next: Node<T> | null = null;
  public prev: Node<T> | null = null;
  constructor(public value: T) {}
}

interface ILinkedList<T> {
  insertAtBeginning: (value: T) => Node<T>;
  insertAtEnd: (value: T) => Node<T>;
  removeAtBeginning: () => void;
  removeAtEnd: () => void;
  insertAtIndex: (value: T, index: number) => Node<T> | null;
  removeAtIndex: (index: number) => void;
  getSize: () => number;
  getHead: () => Node<T> | null;
  getTail: () => Node<T> | null;
  getAtIndex: (index: number) => Node<T> | null;
  toArray: () => Array<T>;
}

export class LinkedList<T> implements ILinkedList<T> {
  private head: Node<T> | null = null;
  private size: number = 0;

  insertAtBeginning(value: T) {
    const node = new Node(value);
    if (this.head) {
      this.head.prev = node;
      node.next = this.head;
      this.head = node;
    } else {
      this.head = node;
    }
    this.size++;
    return node;
  }

  insertAtEnd(value: T) {
    const node = new Node(value);
    const tail = this.getTail();
    if (tail) {
      node.prev = tail;
      tail.next = node;
    } else {
      this.head = node;
    }
    this.size++;
    return node;
  }

  getHead(): Node<T> | null {
    return this.head;
  }

  getTail(): Node<T> | null {
    if (!this.head) {
      return null;
    } else {
      const getLast = (node: Node<T>): Node<T> => {
        return node.next ? getLast(node.next) : node;
      };
      return getLast(this.head);
    }
  }

  getAtIndex(index: number): Node<T> | null {
    if (!this.head) {
      return null;
    }

    const travelTo = (node: Node<T>, index: number): Node<T> | null => {
      console.log(`travelling: index ${index}, value ${node.value}`);
      if (index === 0) {
        console.log("reached the index");
        return node;
      }
      if (!node.next) {
        console.log("no more nodes, returning null");
        return null;
      }
      console.log("lets go to next");
      return travelTo(node.next, index - 1);
    };

    return travelTo(this.head, index);
  }

  getSize() {
    return this.size;
  }

  deleteNode(node: Node<T>): void {
    if (!node.prev) {
      console.log("delete node: this node is head!");
      this.head = node.next;
      if (this.head) {
        this.head.prev = null;
      }
    } else {
      console.log("delete node: this node is not head.");
      console.log(`node prev: ${node.prev.value}`);
      node.prev.next = node.next;
      if (node.next) {
        node.next.prev = node.prev;
      }
    }
    this.size--;
  }

  removeAtBeginning() {
    if (!this.head) {
      console.log("no head from list itself!");
      return;
    }
    console.log("deleting from head");
    this.deleteNode(this.head);
  }

  removeAtEnd() {
    const tail = this.getTail();
    if (!tail) {
      return;
    }
    this.deleteNode(tail);
  }

  insertAtIndex(value: T, index: number) {
    const node = this.getAtIndex(index);
    if (!node) {
      console.log("node of such index does not exist!");
      return null;
    }
    const replacement = new Node(value);
    if (node.prev) {
      console.log(`fixing up previous node: ${node.prev.value}`);
      node.prev.next = replacement;
      replacement.prev = node.prev;
      console.log(
        `previous node's new next neighbour: ${node.prev.next.value}`
      );
    } else {
      replacement.prev = null;
      console.log(
        "current occupant does not have a previous node, it must be head!"
      );
    }
    replacement.next = node;
    console.log(
      `and the next node would be the past occupant: ${replacement.next.value}`
    );
    node.prev = replacement;
    console.log(
      `fixing up current node's previous neighbour: ${node.prev.value}`
    );
    return replacement;
  }

  removeAtIndex(index: number) {
    const node = this.getAtIndex(index);
    if (!node) {
      return;
    }
    this.deleteNode(node);
  }

  toArray(): Array<T> {
    const array: Array<T> = [];
    if (!this.head) {
      return array;
    }

    const addToArray = (node: Node<T>): Array<T> => {
      array.push(node.value);
      return node.next ? addToArray(node.next) : array;
    };

    return addToArray(this.head);
  }
}
