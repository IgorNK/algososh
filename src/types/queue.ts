export interface IQueueDisplayHandler {
  enqueue: (item: string) => void;
  dequeue: () => void;
  clearQueue: () => void;
}

interface IQueue<T> {
  enqueue: (item: T) => void;
  dequeue: () => T | null;
  peek: () => T | null;
  copy: () => IQueue<T>;
  isEmpty: () => boolean;
  getHead: () => number;
  getTail: () => number;
  toArray: () => (T | null)[];
}

export class Queue<T> implements IQueue<T> {
  private items: (T | null)[] = [];
  private head = 0;
  private tail = -1;
  private readonly size: number = 0;
  private length: number = 0;

  constructor(size: number) {
    console.log(`creating queue with size: ${size}`);
    this.size = size;
    this.items = Array(size);
    this.tail = size - 1;
  }

  enqueue(item: T) {
    console.log(`length: ${this.length}, size: ${this.size}`);
    if (this.length >= this.size) {
      throw new Error("Maximum length exceeded");
    }
    const position = (this.tail + 1) % this.size;
    console.log(`queueing in position: ${position}`);
    this.items[position] = item;
    this.tail = position;
    this.length++;
    console.log(
      `head: ${this.head}, tail: ${this.tail}, length: ${this.length}`,
    );
  }

  dequeue() {
    if (this.isEmpty()) {
      throw new Error("No elements in the queue");
    }

    const item = this.items[this.head];
    this.items[this.head] = null;
    this.head = (this.head + 1) % this.size;
    console.log(`this.head + 1: ${this.head + 1}, size: ${this.size}`);
    this.length--;
    console.log(`head: ${this.head}, tail: ${this.tail}`);
    return item;
  }

  peek() {
    if (this.isEmpty()) {
      throw new Error("No elements in the queue");
    }
    return this.items[this.head];
  }

  isEmpty() {
    return this.length === 0;
  }

  getHead() {
    return this.head;
  }

  getTail() {
    return this.tail;
  }

  copy() {
    const newQueue = new Queue<T>(this.size);
    newQueue.items = this.items.slice();
    newQueue.head = this.head;
    newQueue.tail = this.tail;
    newQueue.length = this.length;
    return newQueue;
  }

  toArray() {
    const newArray = [];
    for (let i = 0; i < this.size; i++) {
      const element = this.items[i];
      if (element) {
        newArray.push(element);
      } else {
        newArray.push(null);
      }
    }
    return newArray;
  }
}
