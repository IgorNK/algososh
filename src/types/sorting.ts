export enum SortMethod {
  Selection,
  Bubble,
}
