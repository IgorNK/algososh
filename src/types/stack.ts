export interface IStackDisplayHandler {
  addToStack: (item: string) => void;
  popFromStack: () => void;
  clearStack: () => void;
}

interface IStack<T> {
  push: (item: T) => void;
  pop: () => T | null;
  peek: () => T | null;
  getSize: () => number;
  isEmpty: () => boolean;
  copy: () => Stack<T>;
  toArray: () => Array<T>;
}

export class Stack<T> implements IStack<T> {
  private items: T[] = [];

  constructor() {
    console.log("creating new stack");
  }

  push(item: T) {
    this.items.push(item);
  }

  pop() {
    const item = this.items.pop();
    if (item !== undefined) {
      return item;
    }
    return null;
  }

  peek() {
    if (this.isEmpty()) {
      return null;
    }
    return this.items[this.getSize() - 1];
  }

  getSize() {
    return this.items.length;
  }

  isEmpty() {
    return this.getSize() === 0;
  }

  copy() {
    const newStack = new Stack<T>();
    newStack.items = this.items.slice();
    return newStack;
  }

  toArray() {
    return this.items.slice();
  }
}
