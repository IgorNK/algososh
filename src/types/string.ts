export interface IStringReverseHandler {
  reverse: () => Promise<void>;
}
