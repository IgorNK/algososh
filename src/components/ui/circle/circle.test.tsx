import React from "react";
import renderer from "react-test-renderer";
import { render, screen, fireEvent } from "@testing-library/react";
import { Circle } from "./circle";
import { ElementStates } from "../../../types/element-states";

const TestElement: React.FC = (props) => {
  return <div id="Test Element"></div>;
};

it("Circle without letters is rendered correctly", () => {
  const tree = renderer.create(<Circle />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("Circle with letters is rendered correctly", () => {
  const tree = renderer.create(<Circle letter="A" />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("Circle with head is rendered correctly", () => {
  const tree = renderer.create(<Circle head="A" />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("Circle with JSX-element in head is rendered correctly", () => {
  const tree = renderer.create(<Circle head={<TestElement />} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("Circle with tail is rendered correctly", () => {
  const tree = renderer.create(<Circle tail="A" />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("Circle with JSX-element in tail is rendered correctly", () => {
  const tree = renderer.create(<Circle tail={<TestElement />} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("Circle with prop isSmall === true is rendered correctly", () => {
  const tree = renderer.create(<Circle isSmall={true} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("Circle with state prop set as Default is rendered correctly", () => {
  const tree = renderer
    .create(<Circle state={ElementStates.Default} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("Circle with state prop set as Changing is rendered correctly", () => {
  const tree = renderer
    .create(<Circle state={ElementStates.Changing} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("Circle with state prop set as Modified is rendered correctly", () => {
  const tree = renderer
    .create(<Circle state={ElementStates.Modified} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
