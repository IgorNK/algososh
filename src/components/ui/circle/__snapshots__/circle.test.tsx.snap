// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circle with JSX-element in head is rendered correctly 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  >
    <div
      id="Test Element"
    />
  </div>
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Circle with JSX-element in tail is rendered correctly 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  >
    <div
      id="Test Element"
    />
  </div>
</div>
`;

exports[`Circle with head is rendered correctly 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head string"
  >
    A
  </div>
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Circle with letters is rendered correctly 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    >
      A
    </p>
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Circle with prop isSmall === true is rendered correctly 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle  small default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Circle with state prop set as Changing is rendered correctly 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   changing"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Circle with state prop set as Default is rendered correctly 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Circle with state prop set as Modified is rendered correctly 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   modified"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Circle with tail is rendered correctly 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 string"
  >
    A
  </div>
</div>
`;

exports[`Circle without letters is rendered correctly 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;
